src/python/grpcio_tests/tests/fork/methods.py

```diff
@@ -118,33 +124,131 @@ def __init__(self, task, args=None):
+        cmd = [
+            "gdb",
+            "-ex",
+            "set confirm off",
+            "-ex",
+            "attach {}".format(os.getpid()),
+            "-ex",
+            "set follow-fork-mode child",
+            "-ex",
+            "continue",
+            "-ex",
+            "bt",
+        ]
+        streams = tuple(tempfile.TemporaryFile() for _ in range(2))
+        sys.stderr.write("Invoking gdb\n")
+        sys.stderr.flush()
+        process = subprocess.Popen(cmd, stdout=sys.stderr, stderr=sys.stderr)
+        time.sleep(5)
 
     def start(self):
-        self._process.start()
+        import sys
+        sys.stderr.write("AAAAAAAAAAAAAAAAAAAAAA forking\n")
+
+        # NOTE: Try uncommenting the following line if the child is segfaulting.
+        # self._orchestrate_child_gdb()
+        ret = os.fork()
+        if ret == 0:
+            self._child_main()
+        else:
+            self._child_pid = ret
+
+    def wait(self, timeout):
+        total = 0.0
+        wait_interval = 1.0
+        while total < timeout:
+            ret, termination = os.waitpid(self._child_pid, os.WNOHANG)
+            if ret == self._child_pid:
+                self._rc = os.waitstatus_to_exitcode(termination)
+                return True
+            time.sleep(wait_interval)
+            total += wait_interval
+        else:
+            return False
+
+    def _print_backtraces(self):
+        cmd = [
+            "gdb",
+            "-ex",
+            "set confirm off",
+            "-ex",
+            "echo attaching",
+            "-ex",
+            "attach {}".format(self._child_pid),
+            "-ex",
+            "echo print_backtrace",
+            "-ex",
+            "thread apply all bt",
+            "-ex",
+            "echo printed_backtrace",
+            "-ex",
+            "quit",
+        ]
+        streams = tuple(tempfile.TemporaryFile() for _ in range(2))
+        sys.stderr.write("Invoking gdb\n")
+        sys.stderr.flush()
+        process = subprocess.Popen(cmd, stdout=streams[0], stderr=streams[1])
+        try:
+            process.wait(timeout=_GDB_TIMEOUT_S)
+        except subprocess.TimeoutExpired:
+            sys.stderr.write("gdb stacktrace generation timed out.\n")
+        finally:
+            for stream_name, stream in zip(("STDOUT", "STDERR"), streams):
+                stream.seek(0)
+                sys.stderr.write("gdb {}:\n{}\n".format(
+                    stream_name,
+                    stream.read().decode("ascii")))
+                stream.close()
+            sys.stderr.flush()
 
     def finish(self):
-        self._process.join(timeout=_CHILD_FINISH_TIMEOUT_S)
-        if self._process.is_alive():
-            raise RuntimeError('Child process did not terminate')
-        if self._process.exitcode != 0:
-            raise ValueError('Child process failed with exitcode %d' %
-                             self._process.exitcode)
+        import sys
```

Comment:
Are this and below lines intended? Looks like debug strings to me.

Reply:
You're right. Probably not necessary anymore.